### PR TITLE
op-program: Prune L1 blocks after advancing origin

### PR DESCRIPTION
**Pull request by ajsutton**
**Description**

Prune the L1 blocks array in batch queue after advancing origin since the advancement may have added the parent L1 origin which is required to prune successfully. Previously if the initial L1 block used as origin was exactly 1 block prior to the L1 origin of the safe head, the batch queue broke its invariant for `l1Blocks` field and failed to pop off the initial L1 block because the safe head L1 origin wasn't in `l1Blocks` when the trim happened but was then added as part of advancing the origin (indicating that all L2 blocks with the initial L1 block as their origin were already safe).

Add e2e test to reproduce corner case when deploying a chain at L1 genesis.

**Tests**

Added unit test and action test.

**Additional context**

Fixes flaky test `TestPrecompiles/DisputePrecompile-ecrecover`

### CODE REVIEW

File: op-node/rollup/derive/batch_queue.go [hunk 1]
```diff
@@ -146,6 +131,22 @@ func (bq *BatchQueue) NextBatch(ctx context.Context, parent eth.L2BlockRef) (*Si
 		bq.log.Info("Advancing bq origin", "origin", bq.origin, "originBehind", originBehind)
 	}
 
+	// If the epoch is advanced, update bq.l1Blocks
+	// Advancing epoch must be done after the pipeline successfully apply the entire span batch to the chain.
+	// Because the span batch can be reverted during processing the batch, then we must preserve existing l1Blocks
+	// to verify the epochs of the next candidate batch.
+	// It also
```
Comment (sebastianst) [inline op-node/rollup/derive/batch_queue.go]: sentence?

File: op-e2e/e2eutils/disputegame/claim_helper.go [hunk 2]
```diff
@@ -98,6 +99,10 @@ func (c *ClaimHelper) RequireCorrectOutputRoot(ctx context.Context) {
 	c.require.Equalf(expected, c.claim, "Should have correct output root in claim %v and position %v", c.Index, c.Position)
 }
 
+func (c *ClaimHelper) RequireInvalidStatusCode() {
+	c.require.Equal(byte(mipsevm.VMStatusInvalid), c.claim[0], "should have had valid status code")
```
Comment (sebastianst) [inline op-e2e/e2eutils/disputegame/claim_helper.go]:
```suggestion
	c.require.Equal(byte(mipsevm.VMStatusInvalid), c.claim[0], "should have had invalid status code")
```
?

File: op-e2e/actions/batch_queue_test.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	require.EqualValues(gt, l2BlockNum, seqEngine.l2Chain.CurrentFinalBlock().Number.Uint64())
+
+	// Create a new verifier using the existing engine so it already has the safe and finalized heads set.
+	// This is the same situation as if op-node restarted at this point.
+	l2Cl, err := sources.NewEngineClient(seqEngine.RPCClient(), logger, nil, sources.EngineClientDefaultConfig(sd.RollupCfg))
+	require.NoError(gt, err)
+	verifier := NewL2Verifier(t, logger, sequencer.l1, miner.BlobStore(), altda.Disabled, l2Cl, sequencer.rollupCfg, sequencer.syncCfg, safedb.Disabled)
+	verifier.ActL2PipelineFull(t) // Should not get stuck in a reset loop forever
+	require.EqualValues(gt, l2BlockNum, seqEngine.l2Chain.CurrentSafeBlock().Number.Uint64())
+	require.EqualValues(gt, l2BlockNum, seqEngine.l2Chain.CurrentFinalBlock().Number.Uint64())
```
Comment (protolambda) [inline op-e2e/actions/batch_queue_test.go:94]: Did you mean to assert the sync-status of the verifier op-node?
The sequencer engine is shared, and the expected `l2BlockNum` did not change.

**Review by protolambda — COMMENTED** [review]
LGTM, one question about a test assertion that seems off though

File: op-e2e/actions/batch_queue_test.go (hunk 3 above)
Comment (ajsutton) [reply op-e2e/actions/batch_queue_test.go:94]: The fail case is actually that the `verifier.ActL2PipelineFull(t)` calls gets stuck in an infinite loop of resetting so this was just a sanity check that it hadn't changed things, but yeah it would make more sense to check the sync status of op-node. Will adjust.

**Review by protolambda — APPROVED** [review]
